src/consumption.cpp

```diff
@@ -300,9 +317,17 @@ morale_type player::allergy_type( const item &food ) const
     };
 
     for( const auto &tp : allergy_tuples ) {
-        if( has_trait( std::get<0>( tp ) ) &&
-            food.has_flag( std::get<1>( tp ) ) ) {
-            return std::get<2>( tp );
+        if( food.components.empty() ) {
+            if( has_trait( std::get<0>( tp ) ) &&
+                food.has_flag( std::get<1>( tp ) ) ) {
+                return std::get<2>( tp );
+            }
+        }
+        for( auto comp : food.components ) {
```

Comment:
This shouldn't require a copy of the item in the `components` vector, a const reference would be good.